domain/optatask/src/main/java/org/optatask/domain/realtime/PinTaskProblemFactChange.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optatask.domain.realtime;
+
+import com.thoughtworks.xstream.annotations.XStreamAlias;
+import org.optaplanner.core.impl.score.director.ScoreDirector;
+import org.optaplanner.core.impl.solver.ProblemFactChange;
+import org.optatask.domain.AbstractPersistable;
+import org.optatask.domain.Task;
+import org.optatask.domain.TaskAssigningSolution;
+
+@XStreamAlias("TaPinTaskProblemFactChange")
+public class PinTaskProblemFactChange extends AbstractPersistable implements ProblemFactChange<TaskAssigningSolution> {
+
+    private int consumedTime;
+
+    public PinTaskProblemFactChange(int consumedTime) {
+        this.consumedTime = consumedTime;
+    }
+
+    @Override
+    public void doChange(ScoreDirector<TaskAssigningSolution> scoreDirector) {
+        scoreDirector.getWorkingSolution().setFrozenCutoff(consumedTime);
+
+        for (Task workingTask : scoreDirector.getWorkingSolution().getTaskList()) {
+            if (workingTask.getStartTime() != null && workingTask.getStartTime() < consumedTime) {
```

Comment:
I wonder if we should first call:
```java
workingTask = scoreDirector.lookUpWorkingObject(workingTask);
```
Really don't know if it's necessary or not.

Reply:
I discussed this with @ge0ffrey the other day, scoreDirector.lookUpWorkingObject() is a faster method to find an object given it's id than looping the the whole list in the working solution to find it. So if we're looping through the whole list anyway it is unnecessary (and adds more overhead) to call it again.

Reply:
Thanks, it's worth leaving an explaining comment in the code. Will be useful for future readers.